File: trap.py
class Solution:
    '''
    Leetcode Solution - most optimal - a new type to rememeber 
    '''
    def trap(self, height: List[int]) -> int:
        if not height or len(height) == 0:
            return 0
        lw , rw = 0, 0
        l , r = 0 , len(height)-1 
        result = 0 
        while l <=  r:
            if lw <= rw:
                if lw > height[l]:
                    result = result + lw - height[l]
                else:
                    lw = height[l]
                l+= 1
            else:
                if rw > height[r]:
                    result = result + rw - height[r]
                else:
                    rw = height[r]
                r -= 1 
        return result

File: trap.java
//TC - O(n);
//SC - O(1);
//LC - 42;
class Solution {
    public int trap(int[] height) {
        
        int len = height.length;
        
        if(len==0 || len==1 || len==2) return 0;
        
        int l = 0;
        int r = len-1;
        
        int lw=0,rw=0;
        int result=0;
        
        while(l<=r){
            if(lw<=rw){
                if(height[l]<lw){
                    result += lw-height[l];
                }
                else{
                    lw=height[l];
                }
                l++;
            }
            else{
                if(height[r]<rw){
                    result += rw-height[r];
                }
                else{
                    rw=height[r];
                }
                r--;
                
            }
        }
        
        return result;
        
    }
}